### DISCUSSION TITLE
Lesson 9. Interaction.s.sol

### DISCUSSION THREAD

**Discussion by ducanh2706** [Cyfrin/foundry-full-course-cu#1032] [General]
https://github.com/Cyfrin/foundry-smart-contract-lottery-f23/blob/main/script/Interactions.s.sol

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Script, console} from "forge-std/Script.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {Raffle} from "../src/Raffle.sol";
import {DevOpsTools} from "foundry-devops/src/DevOpsTools.sol";
import {VRFCoordinatorV2Mock} from "../test/mocks/VRFCoordinatorV2Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";

contract CreateSubscription is Script {
    function createSubscriptionUsingConfig() public returns (uint64) {
        HelperConfig helperConfig = new HelperConfig();
        (
            ,
            ,
            ,
            ,
            ,
            address vrfCoordinatorV2,
            ,
            uint256 deployerKey
        ) = helperConfig.activeNetworkConfig();
        return createSubscription(vrfCoordinatorV2, deployerKey);
    }

    function createSubscription(
        address vrfCoordinatorV2,
        uint256 deployerKey
    ) public returns (uint64) {
        console.log("Creating subscription on chainId: ", block.chainid);
        vm.startBroadcast(deployerKey);
        uint64 subId = VRFCoordinatorV2Mock(vrfCoordinatorV2)
            .createSubscription();
        vm.stopBroadcast();
        console.log("Your subscription Id is: ", subId);
        console.log("Please update the subscriptionId in HelperConfig.s.sol");
        return subId;
    }

    function run() external returns (uint64) {
        return createSubscriptionUsingConfig();
    }
}
```
I don't understand why we have to use VRFCoordinatorV2Mock in here instead of the real interface VRFCoordinatorV2Interface. Is the VRFCoordinatorV2Mock compatible with the contract when I deploy on the Sepolia testnet/mainet?

**Comment by theirrationalone** [comment]
Actually we don't want overhead to create  a strong protective 101% randomness to announce a winner. We just need a way to see & interact with chainlink's potential that's why we have, Actually Sir Patrick! created a logical SMC to play with chainlink's randomness Guarantee....

by far.... You can make a dive into that VRF Coordinator to understand things better. There's just some chisel's marks using some `solidity's globals` those you'll find Beneficial into Course's challenges.

However, if you wanna implement real vrfcoordinator into your Raffle SMC then you can do so, nobody is stopping you to do that just read its docs import vrfcoordinatorv2interface and start implementing.

;`)

**Reply by ducanh2706** [reply]
Actually I think you misunderstood my question. What I mean is if we have the interface, we just need to pass the coordinator address to call it right? why use the mock (because I think the mock is implemented very different from the real chainlink vrf)? The code use the mock implementation to pass the vrf coordinator address (in this circumstance, I am using the ethereum mainet)?

Sorry if I said anything wrong (I'm a newbie) ❤️

**Reply by ducanh2706** [reply]
As far as i know the mock is mainly for testing

**Reply by theirrationalone** [reply]
Sorry for that 😅,

Because we've used `mock` implementation and you know what that mock is???? It's just a SMC(smart contract) as typical as we have other SMCs. It's not an interface. Since we mocked up the vrfcoordinator that's why we're not passing a vrfcoordinator address therefore we've our own way, own mocked vrfcoordinator to deal with our intended logic.

And the main thing that should be cleared....

We're using vrfcoordinator in our `raffle` or `lottery`. So when our `raffle` shall be on the mainnet... it'll use real vrfcoordinator. Remember we have mocked up that very same vrfcoordinator for testing purposes only.

Notice carefully, When we're deploying our `raffle` contract locally on temporary anvil chain spun by foundry, we're using our `mock` right.

Similarly, When when we deploy our `raffle` on a local testnet i.e., rinkeby, goerli, etc, we pass a respective `vrfcoordnator` address (look into helperConfig). After deployment `raffle` uses that `vrfcoordinator` address to announce our winner (Hopefully contract associated with `vrfcoordinator` address is already available on that testnet chain.

Chronologically..... When we deploy our `raffle` on a mainnet i.e., ethereum, polygon, etc... Contract associated with `vrfcoordinator` address has its existence on to the mainnet too. Look into `helperConfig` for this too, On there, we also have used `vrfcoordinator` address variant that's available on mainnet respectively.

**Reply by ducanh2706** [reply]
yeh thanks for your detailed explanation. But I still don't understand what's going on. I mean, in this code:

```solidity
vm.startBroadcast(deployerKey);
      uint64 subId = VRFCoordinatorV2Mock(vrfCoordinatorV2)
          .createSubscription();
      vm.stopBroadcast();
```
What if I deploy on sepolia testnet? is the VRFCoordinatorV2Mock logic compatible with the contract at address of vrfCoordinatorV2?

**Reply by theirrationalone** [reply]
Naah, never, if it could be possible then again we would be on kind of centralization.... 

It could be similar but can't be compatible.... It would mean that you're(contract) pretending to be the chainlink vrfcoordinatorV2. It's okay to be a mockingbird for testing purposes only but never for testnets Or mainnets.

Yes for same codebase with different deployment addresses but who will vote for it... Majority will rule and they will kick that contract out. 

Mantra:

Anything Should be possible locally for testing purposes only. 

Nothing Should be capable of to be a pretender Or a goon. 



You can deploy your logic anywhere on blockchain world but eventually will majority will come (He who remains ;`) ..... 

Think of LOKI.... 

There could be your variants..... If they Or even you try to violate sacred timeline.... TVA will come to the rescue... To defend against infinite variants.... 

BLOCKCHAIN === LOKI WEB SERIES. 

Anonymous (satoshi nakamoto) aka TVA | Sacred timeline creator | maintainer === The Kang. 

Adversary | Attacker === Loki character Or anybody else Who ever try to violate sacred timeline. 

If branches start emerging then infinite Kang's variants can come for destruction.... 


Now watch and think.... 

Too comical haaan 😅😅

;`)

**Reply by ducanh2706** [reply]
Thanks for your enthusiasm!!!! I think I do understand it now. So if i use this code for sepolia/main net:

```solidity
vm.startBroadcast(deployerKey);
      uint64 subId = VRFCoordinatorV2Mock(vrfCoordinatorV2)
          .createSubscription();
      vm.stopBroadcast();
```
the createSubscription() function of the mock will actually override the actual createSubscription() of vrfCoordinatorV2 on the sepolia/main net right? (I have tried to deploy this code on sepolia testnet and it still worked)
 Is it the right solution to fix the above solution?
 ```solidity
vm.startBroadcast(deployerKey);
      uint64 subId = VRFCoordinatorV2Interface(vrfCoordinatorV2)
          .createSubscription();
      vm.stopBroadcast();
```

**Reply by theirrationalone** [reply]
Yes ofcourse buddy! 

I think you should consider interfaces in solidity.. Right now.


May this explanation help you to understand that better..... 


Interface: i don't care about any provided addresses.... If the contract associated with a given address has very same functions those i have... Then i will definitely execute them whenever i found myself calling them.

**Reply by ducanh2706** [reply]
Thank you so much ❤️❤️❤️

**Reply by theirrationalone** [reply]
You're welcome @ducanh2706 !  ❤

If you are still facing hard time to get that... It might be me going wrong ..... Btw i know some of community members who can help you further.... Should i call them here????
